client/consensus/manual-seal/src/heartbeat_stream.rs

```diff
@@ -0,0 +1,100 @@
+// Library shortcuts
+
+use std::{pin::Pin};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
+
+use crate::rpc::{EngineCommand};
+
+// ---
+// Constant Definition
+
+const DEFAULT_TIMEOUT: u64 = 30;
+const DEFAULT_MIN_BLOCKTIME: u64 = 1;
+const DEFAULT_FINALIZE: bool = false;
+
+// ---
+pub struct HeartbeatOptions {
+	// the amount of time passed that a new heartbeat block will be generated, in sec.
+	pub timeout: u64,
+	// the minimum amount of time to pass before generating another block, in sec.
+	pub min_blocktime: u64,
+	// whether the generated heartbeat block is finalized
+	pub finalize: bool,
+}
+
+impl Default for HeartbeatOptions {
+	fn default() -> Self {
+		Self {
+			timeout: DEFAULT_TIMEOUT,
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+pub struct HeartbeatStream<Hash> {
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	delay: Delay,
+	instant: Instant,
+	opts: HeartbeatOptions,
+}
+
+impl<Hash> HeartbeatStream<Hash> {
+	pub fn new(
+		pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+		opts: HeartbeatOptions
+	) -> Self {
+		Self {
+			pool_stream,
+			delay: tokio::time::delay_for(Duration::from_secs(opts.timeout)),
+			instant: Instant::now(),
+			opts
+		}
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		let mut hbs = self.get_mut();
+		match hbs.pool_stream.poll_next_unpin(cx) {
+			Poll::Ready(Some(ec)) => {
+				// Ensure we are not producing block too frequently, based on `min_blocktime` setting
+				if Instant::now() - hbs.instant < Duration::from_secs(hbs.opts.min_blocktime) {
+					return Poll::Pending;
+				}
```

Comment:
You've checked to make sure we don't author two blocks in short succession, but you haven't ensured that a second block will get authored in a reasonable amount of time. Imagine a node using instant seal. A user submits two transactions very close together. The first transaction causes a block to be authored. The second transaction does not cause a block to be authored because of this logic.

But that means the second transaction will sit in the pool until the next heartbeat block comes which will generally be a long time or possibly never if heartbeats are disabled. I think better behavior is that when a transaction arrives and it is too soon to author another block, a timer should be started so that the next block will be authored as soon the minimum block time runs out.